### ISSUE TITLE
Slow access to ome-zarr at lower levels

### ISSUE THREAD

**Issue by pwinston** [napari/napari#2012]
## 🐛 Bug

For a while, I've observed when I zoom into `4495402.zarr` off SSD with octree it gets slower the more you zoom. I assumed it was something with the octree code. But these traces show it's the actual data loads that get slower. Here we are viewing a high level of the octree. This read was 12.4ms and it was one of the slower ones:

<img width="901" alt="zarr-fast" src="https://user-images.githubusercontent.com/4163446/102027307-3681b100-3d71-11eb-8a71-6c972d6a5583.png">

But here's level 0 and now the loads are all over 150ms:

<img width="936" alt="zarr-slow" src="https://user-images.githubusercontent.com/4163446/102027314-439ea000-3d71-11eb-9ccf-6b9f7b83138f.png">

It's very consistent too. You never see a slow read on a high level or a fast read on a level 0. Above notice that the draws have gotten slower in tandem with the reads getting slower. This is likely due to the GIL. One of these threads is slowing the other one down. But you can't tell which is which. Are the reads making the rendering slow, or vice versa?

To see, I call this test routine when I create the slice:

    def _test_speed(data):
        for i, level in enumerate(reversed(data)):
            with block_timer(f"level {len(data) - 1- i} asarray", print_time=True):
                np.asarray(level[0, 0])

Loading from an external SSD I get this output:

     level 7 asarray 6.500ms
     level 6 asarray 3.534ms
     level 5 asarray 3.649ms
     level 4 asarray 3.989ms
     level 3 asarray 5.988ms
     level 2 asarray 12.994ms
     level 1 asarray 49.613ms
     level 0 asarray 189.473ms

So this implies the reads are causing the rendering to be slow, not the reverse. It implies to me these reads are doing a ton of CPU work, and that's slowing down the rendering. Although note even if the rendering was fast, the loads would still be happening slowly. That is, we'd have a high framerate but we'd be watching the tiles slowly pop in.

So just calling `np.asarray` to get one pixel takes 189ms on level 0. I don't know if the issue here is OME, Dask, Zarr or something else? However, we saw something like this with Dask a while back. Where something ran much slower then expected. Note that level 0 here is huge:

    380928 x 921600 = 351.1 billion pixels.

It could the number of Dask chunks is really big, and there's some cost that proportional to the number of chunks? If so should we be using larger chunks? Or more than one level of chunks?

We need the load time to be a lot faster if we are going to render it quickly. The ping time from US to Europe is around 150ms which is what we are getting from the SSD here.

**Comment by d-v-b** [comment]
what's the chunk size of each level? zarr will at minimum read + decompress a full chunk, so if the chunk size is varying by level this might be a simple explanation.

**Comment by pwinston** [comment]
All levels use 1024 chunks. I don't know a concise way to determine that but just printing `level.chunks` it's just tons of 1024 values and every level is the same. Like level 0 is:

    level 0 chunks = ((1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024,
    1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024,
    1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024,
    1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024,
    1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, 1024, ...

**Comment by pwinston** [comment]
I notice the directory `4495402.zarr/0` contains `334800` files. When we created a chunked file format at my last job, we quickly realized that having more than 10,000 files in a directory created problems. So our format had subdirectories and only 1000 files in each subdirectory. That may have nothing to do with the slowness here, but that's an issue we ran into and that's how we solved it.

I can write a small program that just opens files in each of these directories. If that alone were slow, then napari/dask/zarr would have no hope of opening them fast.

**Comment by d-v-b** [comment]
this behavior (putting all the chunks in one folder) is a bad default for zarr. Try using the `NestedDirectoryStore` (which makes a hierarchy of directories) instead of the vanilla `DirectoryStore` (what zarr uses by default)

**Comment by pwinston** [comment]
I'm not creating the zarr file, just reading it.

I tested just reading one chunk at each level it they were equally fast. First clearing the disk cache with `purge`:

    Read /data-ext/4495402.zarr/6/0.0.0.0.0 in 3.500ms
    Read /data-ext/4495402.zarr/5/0.0.0.0.0 in 0.280ms
    Read /data-ext/4495402.zarr/4/0.0.0.0.0 in 0.336ms
    Read /data-ext/4495402.zarr/3/0.0.0.0.0 in 0.284ms
    Read /data-ext/4495402.zarr/2/0.0.0.0.0 in 0.285ms
    Read /data-ext/4495402.zarr/1/0.0.0.0.0 in 0.251ms
    Read /data-ext/4495402.zarr/0/0.0.0.0.0 in 0.356ms

Then a second run while cached:

    Read /data-ext/4495402.zarr/6/0.0.0.0.0 in 0.093ms
    Read /data-ext/4495402.zarr/5/0.0.0.0.0 in 0.044ms
    Read /data-ext/4495402.zarr/4/0.0.0.0.0 in 0.036ms
    Read /data-ext/4495402.zarr/3/0.0.0.0.0 in 0.033ms
    Read /data-ext/4495402.zarr/2/0.0.0.0.0 in 0.032ms
    Read /data-ext/4495402.zarr/1/0.0.0.0.0 in 0.033ms
    Read /data-ext/4495402.zarr/0/0.0.0.0.0 in 0.031ms

Then I thought maybe `0.0.0.0.0` was special because it was the first file.

So I read 5 random files and still, all the access was pretty fast after purging:

    Read /data-ext/4495402.zarr/6/0.0.0.2.6 in 2.315ms
    Read /data-ext/4495402.zarr/6/0.0.0.0.13 in 0.507ms
    Read /data-ext/4495402.zarr/6/0.0.0.1.3 in 2.512ms
    Read /data-ext/4495402.zarr/6/0.0.0.0.12 in 0.447ms
    Read /data-ext/4495402.zarr/6/0.0.0.4.5 in 2.059ms
    Read /data-ext/4495402.zarr/5/0.0.0.3.27 in 0.895ms
    Read /data-ext/4495402.zarr/5/0.0.0.7.5 in 2.624ms
    Read /data-ext/4495402.zarr/5/0.0.0.5.8 in 1.501ms
    Read /data-ext/4495402.zarr/5/0.0.0.1.8 in 1.646ms
    Read /data-ext/4495402.zarr/5/0.0.0.1.26 in 0.285ms
    Read /data-ext/4495402.zarr/4/0.0.0.0.47 in 0.348ms
    Read /data-ext/4495402.zarr/4/0.0.0.9.23 in 0.864ms
    Read /data-ext/4495402.zarr/4/0.0.0.0.54 in 0.297ms
    Read /data-ext/4495402.zarr/4/0.0.0.23.45 in 0.308ms
    Read /data-ext/4495402.zarr/4/0.0.0.15.43 in 2.385ms
    Read /data-ext/4495402.zarr/3/0.0.0.33.50 in 2.482ms
    Read /data-ext/4495402.zarr/3/0.0.0.1.107 in 0.441ms
    Read /data-ext/4495402.zarr/3/0.0.0.37.61 in 2.740ms
    Read /data-ext/4495402.zarr/3/0.0.0.25.43 in 1.094ms
    Read /data-ext/4495402.zarr/3/0.0.0.22.62 in 2.725ms
    Read /data-ext/4495402.zarr/2/0.0.0.22.222 in 0.616ms
    Read /data-ext/4495402.zarr/2/0.0.0.29.99 in 2.860ms
    Read /data-ext/4495402.zarr/2/0.0.0.11.119 in 2.378ms
    Read /data-ext/4495402.zarr/2/0.0.0.50.188 in 2.756ms
    Read /data-ext/4495402.zarr/2/0.0.0.83.157 in 0.630ms
    Read /data-ext/4495402.zarr/1/0.0.0.176.354 in 0.591ms
    Read /data-ext/4495402.zarr/1/0.0.0.58.25 in 2.653ms
    Read /data-ext/4495402.zarr/1/0.0.0.181.258 in 0.376ms
    Read /data-ext/4495402.zarr/1/0.0.0.48.406 in 2.790ms
    Read /data-ext/4495402.zarr/1/0.0.0.48.236 in 2.473ms
    Read /data-ext/4495402.zarr/0/0.0.0.328.756 in 0.381ms
    Read /data-ext/4495402.zarr/0/0.0.0.95.719 in 2.761ms
    Read /data-ext/4495402.zarr/0/0.0.0.124.779 in 2.847ms
    Read /data-ext/4495402.zarr/0/0.0.0.201.870 in 0.981ms
    Read /data-ext/4495402.zarr/0/0.0.0.209.7 in 0.267ms

So it's not Python or the filesystem or the drive. But maybe something else is choking on that many files? Maybe `NestedDirectoryStore` is needed for someone else?

**Comment by d-v-b** [comment]
It might be helpful to compare `dask_array(zarr_array)[0,0]` vs `zarr_array[0,0]`. I've previously observed some delays in dask that are O(number of chunks)

**Comment by pwinston** [comment]
When I see them in napari it's already a Dask array. I guess the ome-zarr plugin does that? I can try reading it as zarr in a test program though.

**Comment by d-v-b** [comment]
if dask is the issue, see this issue and particularly this comment: https://github.com/dask/dask/issues/5913#issuecomment-714536079

**Comment by pwinston** [comment]
Running `dask.optimize` on level zero takes 3 seconds, but it doesn't speed up subsequent accesses at all. I wish there was something I could call on the array before and after optimize to see if it did anything? But unless I'm messing something up, optimizing does not seem to help.

**Comment by pwinston** [comment]
I think there are two separate questions here. First, can we make the Dask arrays that the ome-zarr plugin gives us fast? Second, can we make fast Dask arrays ourselves, by any means?

The former would be better. But the latter would good too. So at least we know there's a route. And maybe the ome-zarr plugin can take that route. 

I know very little about Dask. But I can try a few things. I have to assume it can do this. That we are just not using it right.

**Comment by d-v-b** [comment]
If it turns out that dask is the issue and not zarr-python or something else, you could a) try tensorstore as a front-end for zarr arrays and b) reach out to the dask devs about this, since I think they would be very interested in this use case and I know there's some ongoing refactoring of array internals that should offer performance gains.

**Comment by sofroniewn** [comment]
cc @joshmoore who might be interested in this thread

**Comment by pwinston** [comment]
I compared 5 random reads with zarr and dask in a small test program with no napari code.

I constructed the dask array with `da.from_zarr` with all  defaults. I'm not sure what options might be useful here? The chunk size seems to default to 1024. I read 5 locations from  level 7 (smallest) and level 0 (biggest). I ran "purge" between runs, to clear the disk cache. The "file" on disk is 259GB, level 0  is (380928, 921600).

    zarr level 7 - /data-ext/4495402.zarr:
    4.669ms - zarr.convenience.open
    3.881ms - level[7][583, 3784]
    2.671ms - level[7][477, 995]
    3.095ms - level[7][1842, 3451]
    1.630ms - level[7][2695, 5789]
    2.971ms - level[7][908, 5940]
    shape (1, 1, 1, 2976, 7200)

    zarr level 0 - /data-ext/4495402.zarr:
    0.395ms - zarr.convenience.open
    1.401ms - level[0][363332, 513881]
    1.752ms - level[0][358608, 121386]
    3.757ms - level[0][90031, 61591]
    3.134ms - level[0][337152, 380359]
    4.108ms - level[0][115085, 549991]
    shape (1, 1, 1, 380928, 921600)

    dask level 7 - /data-ext/4495402.zarr:
    15.851ms - da.from_zarr
    23.781ms - level[7][2576, 1829]
    6.445ms - level[7][2040, 4236]
    8.730ms - level[7][2520, 3556]
    4.232ms - level[7][2076, 2046]
    5.745ms - level[7][1387, 1120]
    shape (1, 1, 1, 2976, 7200)

    dask level 0 - /data-ext/4495402.zarr:
    506.481ms - da.from_zarr
    90.763ms - level[0][55926, 887621]
    84.647ms - level[0][209603, 491646]
    82.562ms - level[0][184573, 147898]
    89.479ms - level[0][117959, 406597]
    85.062ms - level[0][366724, 632559]
    shape (1, 1, 1, 380928, 921600)

The dask reads are faster than the 150ms I saw in napari. But for level 0 dask ranges from 20X slower to 60X slower than zarr. Obviously zarr is just a container here, while dask has a ton of features. Maybe there are options to make dask faster here? If not it suggests dask might not be the best container for real-time rendering.

Gist with the [test code is here](https://gist.github.com/pwinston/3c4d6e2198500d1f2e7211945a93c651).

**Comment by pwinston** [comment]
I tried passing different chunk sizes to `da.from_zarr` and it worked. Before every test I purged the cache, then did `da.from_zarr` with the **Chunk Size** on the left. Then did 100 random reads and the average is **Read Time** is in milliseconds.

The level 0 shape is (380928, 921600) so for **Chunks Across** I divided 921600 by the chunk size.

Chunk Size | Read Time (ms) | Chunks Across | Total Chunks (approx)
-- | -- | -- | --
1024 | 110.74 | 900 | 810000
2048 | 18.29 | 450 | 202500
4096 | 8.27 | 225 | 50625
8192 | 4.96 | 113 | 12656
16384 | 4.44 | 56 | 3164
32768 | 4.18 | 28 | 791
65536 | 4.00 | 14 | 198
131072 | 4.34 | 7 | 49
262144 | 4.67 | 4 | 12
524288 | 6.03 | 2 | 3
1048576 | 5.00 | 1 | 1

So it goes down quickly. It's not quite as fast as zarr but close enough for rendering.

It clearly doesn't read the full-size chunk's worth of data. Or it would have to read the full dataset for the last one. If you read just one pixel it must only be reading one zarr chunk. So it's like there are two layers of chunks with different chunk sizes.

It could be for what we are doing the Dask chunks don't even help us? We could just have 1 chunk? Maybe the chunks are only for processing and for reading it doesn't matter?

Or else we could choose a chunk size so you have around 16 chunks across, around 256 total chunks? Because that's the low point here when the chunk size is 2**6 = 65536. If we did that as the levels got bigger our chunk size would get bigger, but the number of chunks per level would always be around 256.

I'll see if I can re-chunk it inside napari and test the speed.

Here is the [new Gist](https://gist.github.com/pwinston/b4d7012baf4df40bd39c655bea0fa0f5) with the test code that produced. You have to run it with `sudo` because it runs `purge`.

**Comment by pwinston** [comment]
Very weird results inside napari.

If I rechunk to 65536 it rechunks without error, but then I get weird hard-crashes. I figured I was yanking the data out from under something. So I uninstalled ome-zarr and I'm using the built-in place where we call `da.from_zarr` in `read_zarr_dataset`. So no rechunking, I just pass `da.from_zarr` the chunk size we want.

If I use 2048 chunks I see approximately the speedup above. Which is a pretty nice speedup. But if I make the chunk size any bigger I again get weird crashes right away. And with 2048 it only lasted a few minutes and crashed. But with 1024, what we've been using all long, it's fine.

I just don't see what in our code would know or care about the chunk size. When we access a dask array it's going to return only the bytes we want. Something inside Dask must operate differently. But it's unclear how that could crash us.

This is frustrating because performance is horrible with 1024 chunks.

And it would be totally acceptable with big chunks.

**Comment by pwinston** [comment]
I really have no ideas here! I feel like the chunk size should be a totally internal detail to Dask. I don't see how it's causing crashes in napari. I don't even think we ever SEE the chunk size.

Larger chunks worked fine in the test program. But until we can increase the chunk size in napari we will have this serious performance problem. It will get even worse if the datasets are bigger than this one.

We could switch to using the zarr interface. But that seems silly since Dask has a knob that makes it almost the same speed. We just can't turn the knob!

**Comment by pwinston** [comment]
With 2^16 chunk size, an RGB chunk would be (2^16,  2^16, 4) = 16GB. So maybe dask is allocating memory for the whole chunk. Even though based on the timings above it's definitely not reading the whole chunk into memory, the read takes only 4ms, and the read time doesn't really go up with the chunk size.

[Dask does say](https://docs.dask.org/en/latest/array-chunks.html)  "A chunk should be small enough to fit comfortably in memory. We’ll have many chunks in memory at once". So it could be really big chunks are off the table. However even the 2048 chunk size crashes after a while, and those chunks are only 16MB. So that totally subverts this theory. It seems like can't alter the size of the chunks *at all*. I tried disabling the dask cache, didn't help.

Still, no real ideas on how to fix this. Except use zarr directly, instead of dask's `from_zarr`. Since we can't seem to increase the chunk size without crashes. And the zarr interface has no chunks. It seems to always be fast.

Dask has a million features we aren't using here. We just want to read bytes off disk. What dask is giving us is a consistent interface to access data no matter where the data is coming from. But maybe for zarr files (and others?) we need to bypass that and read the data more directly. That's one idea. But kind of a pain.

**Comment by d-v-b** [comment]
I don't think you want to make your dask chunks much larger than the underlying zarr chunks. And a 16 GB chunksize is a big no-no, because you will blow out memory (which might explain the crashes you are observing). 

What is the chunking of the underlying zarr data? For visualizing 3D data (over http), neuroglancer recommends chunks sized (64,64,64), i.e. a mass of .26 MB for uint8 data. 

If the underlying zarr data has a sensible chunking scheme (~.2 - 1 MB per chunk) AND dask performs poorly when the dask chunks are identical to the zarr chunks, then I think you could raise the issue over in the dask issue tracker. Dask was designed for scheduling computations, not low-latency visualization, so the current implementation of `from_zarr` might be suboptimal for your use case, but I think the devs would be interested in trying to make this work.

**Comment by d-v-b** [comment]
ah i forgot that you already answered the question about the chunksize -- (1024, 1024), so ~1 MB per chunk, which seems reasonable.

**Comment by pwinston** [comment]
Yeah, the zarr file is 1MB/chunk. We know dask does not read a full chunk because with 16GB chunks it takes only 5ms to read a pixel. And that time is basically invariant to chunk size. As long as the chunks are not super small, the reads are basically the same for all chunk sizes.

I'm sure Dask *does* read a full chunk with all its computational pipeline stuff. The stuff it's made for. But for simple reads, it seems to pass that through. Which is good for us. Until we need to do computations I guess.

But I figured out the crashes!!! It's PyQt5 again. We had problems with PyQt5 and "scene graph changes". For instance, our first tiled visual created one VisPy `ImageVisual` per tile. That had constant crashes with PyQt5, but it worked with PySide2.

That's one reason we wrote the new `TiledImageVisual` and `TextureAtlas2D`. Because that doesn't change the scene graph. So that works with PyQt5 and Pyside2.

But Dask chunk size changes also crashes with PyQt5. It does not crash with Pyside2:

| QT_API | Dask Chunks | Behavior |
| -- | -- | -- |
PyQt5 | 1024 | 150ms reads |
PyQt5 | 2048 | crash |
PyQt5 | 65536 | crash |
PySide2  | 1024 | 150ms reads |
PySide2  | 2048 | 30-50ms reads |
PySide2 | 65536 | 4ms reads |

So this is good for me, I can work with this. But it means napari really does need to solve the PyQt5 problem. So scene graph changes *and* dask changes both cause crashes. It could be just "doing stuff with memory" causes the crashes.  Maybe memory on the C++ side? 

I spent a few days looking at the crashes when the scene graph stuff was happening. I didn't figure anything out. It's a very tricky problem. It might be a bug with PyQt5, but it might be we are doing something wrong and PySide2 just happens to not care. We're not sure which. @sofroniewn.

**Comment by pwinston** [comment]
Rechunking doesn't work though. You have to give the larger chunk size to `from_zarr`. Rechunking to 65536 made my reads take 10+ seconds! That might be something to ask Dask about. 

So we can fix our internal zarr reader. But we can only ask plugins to do the same. We can't fix it if they use small chunks. At least I was not able to.

**Comment by d-v-b** [comment]
I think the underlying issue here is the number of chunks. indexing a single element from a dask array created with `from_array` requires O(number of chunks) time, because (I think) there's a step that traverses the entire task graph. In my own tests, indexing time scales linearly with chunk count. So your performance gains from larger chunks are actually due to the commensurate reduction chunk count.

It would be great if dask arrays didn't exhibit this behavior, but I don't know what would need to change internally for indexing to be O(constant).

**Comment by pwinston** [comment]
Yeah, I figured there was O(chunks) in there. Even if they fixed that it might be we need to increase the chunk size as the level gets bigger. We have around 800,000 chunks if set on 1024. But what if the level were 10x or 100x bigger?

I hit kind of the same issue internally. My first version I created kind of an empty tree, storing nothing at each location. But even that was too big for this level 0. So now we create nothing at all, except for what you are currently viewing. So maybe Dask (or some alternative) would need to work the same way to handle trusly immense levels.

**Comment by pwinston** [comment]
So passing a large chunk size to `from_zarr` works with our internal zarr reader and PySide2. And I was able to demo that. But that raises a bunch of questions to fully put this to bed:

1) Addressing the PyQt5 stability problem, the hardest probably.

2) Should we ask for 65536 in our internal `from_zarr`? Or only if using octree? I'm not aware of any downside with large chunks, but there very well might be.

2) Do we need to ask OME and anyone else opening zarr to also do this?

3) Can we get re-chunking working instead? It did not work for me, it made things slower. This would be better because then we "fix" any dask array anyone hands us.

4) Should we just not use Dask if using zarr? Since zarr API is always fast? From day one I felt dask was not made for real-time rendering, they support a lot of use cases, but rendering is just not on their list that I've seen. Maybe we can use Dask for small to medium things but require some type of "raw access" for really huge stuff.

So it's great we found "a solution" but there's still lots of open questions.

**Comment by jni** [comment]
Hi all and thanks for this fascinating discussion! I must admit that

> Yeah, the zarr file is 1MB/chunk. We know dask does not read a full chunk because with 16GB chunks it takes only 5ms to read a pixel. And that time is basically invariant to chunk size. As long as the chunks are not super small, the reads are basically the same for all chunk sizes.

really surprised me and revealed some deep misunderstanding I have about how dask works.

I 100% think the dask folks are interested in improving things for this use case — linear time random access indexing seems like very bad performance in many contexts, not just real time rendering. CC @mrocklin @jakirkham @GenevieveBuckley. In general, to answer your specific questions @pwinston, my intuition is that bigger/massive chunks is a bit of a hack and I would instead aim to get at the root cause of the problem within dask itself — zarr and da.map_blocks show that there's no fundamental reason why smallish chunks should hurt performance quite so much.

In my experience using `da.map_blocks` vs `da.{stack,concatenate}`, it's just night and day. I had a quick look at the [`da.from_zarr`](https://github.com/dask/dask/blob/1d669480f78efa290fad612d196ecd18d26e7972/dask/array/core.py#L3036) and [`da.from_array`](https://github.com/dask/dask/blob/1d669480f78efa290fad612d196ecd18d26e7972/dask/array/core.py#L2861) which is what it resolves to, and I don't know offhand whether it would create one or the other kind of graph. There's probably a way to use `map_blocks` on the output of `zarr.open`. The [dask-image PR for this](https://github.com/dask/dask-image/pull/165/files) is a good model of how to use `map_blocks` generically. There's two other examples in my [rootomics](https://github.com/jni/napari-demos/blob/91b5e8ab03852fc93fe6d74d81fed2d39dbdda3b/rootomics.py#L20-L54) and [napariboard-proto](https://github.com/jni/napariboard-proto/blob/4b00303e567328270e02bd7e977cd2656040f8ea/napariboard.py#L84-L100) demos.

In general, it would be good to get dask working well (a) just for its own sake, but (b) because zarr is generally eager in giving you numpy arrays with indexing, something we've struggled with in the past. But as a stopgap, yes, I will try to prioritise in the next couple of days a PR to ome-zarr-py to improve performance with this use case, either through using zarr.open instead of dask, or through using map_blocks on top of that.

Again, thanks @pwinston for the investigation, definitely taught me some things about dask!

**Comment by pwinston** [comment]
Yes, I agree if `da.from_zarr` could return an interface that had constant-time access regardless of the size of the level, that would be ideal for us. So far we are only reading the data, not modifying it. So maybe we could set a mode or derive a special view of the data that says "suspend everything fancy and just let us read the data as fast as possible."

In the meantime it's great we can solve this temporarily by creating large dask chunks, and that even with huge chunks dask seems to read only the small zarr chunks. That lets me continue to work on the rendering. And I'm just going to assume long term we'll find a way to get that speed for everyone.

I agree large chunks sounds like a hack. It only works because I don't think we do *any* operations that even use the full dask chunks. And that might change. Dask says "A chunk should be small enough to fit comfortably in memory. We’ll have many chunks in memory at once". So we are blatantly violating that advice with 16GB chunks on a 16GB laptop!

**Comment by joshmoore** [comment]
> https://github.com/napari/napari/issues/2012#issuecomment-744141797 @joshmoore who might be interested in this thread

Definitely. Thanks! Interesting read, though I'll need time to dig deeper.

> https://github.com/napari/napari/issues/2012#issuecomment-744097593 a bad default for zarr. Try using the NestedDirectoryStore

Agreed. While rolling out ome-zarr initially there were too many issues trying to support nested+s3 over Python & Java. Looking forward to swapping the default in the new year. Happy to do some testing on the relative performance.

**Comment by pwinston** [comment]
Sounds good. And to be clear @joshmoore one of my test programs above suggested the number of files was not slowing us down. And least I was able to read the files at full speed outside of napari. But it does make using shell pretty slow! And it certainly seems like a bad idea to have everything in one directory at some point, like if you had millions of files.

At one place I worked we had an issue running out of inodes. We had to increase the number. But that might be a problem no matter what the structure is. Just based on the number of files.

**Comment by jakirkham** [comment]
cc @Carreau (who may be interested as well)

**Comment by Carreau** [comment]
> At one place I worked we had an issue running out of inodes. We had to increase the number. But that might be a problem no matter what the structure is. Just based on the number of files.

I have seen this on HPC filesystem (one user mistakenly created 2.5B files across 3 month, one float number per file...), if there are "few enough" files the data is inline in inodes; otherwise you start to have indirection layers, that slow things down.

I'll try to find some time to dig into this; and/or try to see if fix it in spec v3 impl if it's something core to how zarr v2 works.

**Comment by kephale** [comment]
I'm labeling this as wontfix because this part of the codebase has been changed/deprecated. We can revisit/close later.

**Comment by psobolewskiPhD** [comment]
Relevant zulip: https://napari.zulipchat.com/#narrow/channel/212875-general/topic/writing.20to.20zarr.20with.20dask
The fundamental issue here appears to be related to dask and handling of massive number of chunks.

I will say reading that massive array remotely using napari-ome-zarr and new NAPARI_ASYNC=1:
`NAPARI_ASYNC=1 napari https://uk1s3.embassy.ebi.ac.uk/idr/zarr/v0.1/4495402.zarr`
is quite performant!

**Comment by psobolewskiPhD** [comment]
I'm going to close this, since the issue is largely resolved. I think that the napari-ome-zarr plugin may consider using zarr-python and the new CacheStore for better performance for remote data.
Meanwhile, on the napari side we should look into: https://github.com/napari/napari/issues/8976
